### ISSUE TITLE
Not unique table/alias mySql exception with two through relations on another through relation

### ISSUE THREAD

**Issue by sieppl** [yiisoft/yii#1588]
Let's say we have Author, Category, AuthorHasCategory. Each author has a main category he is working on. Therefore the table "AuthorHasCategory" has a boolean "is_main". Category has n:1 relations to Foo and Bar to further describe the Category.

Now we define following relations on Author:

```
public function relations() {
    return array(
            'authorHasCategories' => array(self::HAS_MANY, 'AuthorHasCategory', 'author_id'),
            'mainCategory'=>array(self::HAS_ONE,'Category',array('category_id'=>'id'),'through'=>'authorHasCategories', 'on' => 'is_main = 1'),
            'mainCategoryFoo' => array(self::HAS_ONE, 'Foo', array('foo_id' => 'id'), 'through' => 'mainCategory'),
            //BUG: adding following declaration will not work due to a duplicate table alias exception from MySql
            'mainCategoryBar' => array(self::HAS_ONE, 'Bar', array('bar_id' => 'id'), 'through' => 'mainCategory')
    );
}
```

When mainCategoryFoo and mainCategoryBar are **added to "with" at the same time**, an mysql exception occurs:

```
SQLSTATE[42000]: Syntax error or access violation: 1066 Not unique table/alias: 'authorHasCategories'.
```

When checking the sql query you will find twice:

```
 LEFT OUTER JOIN `author_has_category` `authorHasCategories` ON (`authorHasCategories`.`author_id`=`t`.`id`)
```

The first time for the actual pivot table and another time it seems the whole left other join chain is added another time instead of reuse existing joins.

Sorry for the abstract example, I had to cut it out from our current project.

**Comment by PatchRanger** [comment]
@acorncom Would you mind making a pull request with your fix to main Yii repository? I wish this issue was going to be fixed somehow :)

**Comment by gunzino** [comment]
Hello, I have applied the patch provided by @acorncom 

I have the relations defined this way:

                        'uzivatelZamestnanec' => array(self::HAS_ONE, 'UzivatelZamestnanec', 'uzivatel_id', 'on'=> 'UzivatelZamestnanec.datum_ukoncenia=0'),
                        'pobocka_z'=>array(self::BELONGS_TO,'Pobocka',array('pobocka_id'=>'id'),'through'=>array('uzivatelZamestnanec' =>'uziv_2')),
                        'typ_pracovneho_pomeru'=>array(self::BELONGS_TO,'PracovnyPomerTyp',array('typ_pracovneho_pomeru'=>'id'),'through'=>array('uzivatelZamestnanec' => 'uziv_1')),


But this seems not to work, it stops rendering on this query:
SELECT COUNT(DISTINCT `t`.`id`) FROM `uzivatel` `t` LEFT OUTER JOIN `uzivatel_zamestnanec` `uziv_2` ON (`uziv_2`.`uzivatel_id`=`t`.`id`) AND (datum_ukoncenia=0) LEFT OUTER JOIN `pobocka` `pobocka_z` ON (`pobocka_z`.`id`=`uziv_2`.`pobocka_id`) LEFT OUTER JOIN `uzivatel_zamestnanec` `uziv_2` ON (`uziv_2`.`uzivatel_id`=`t`.`id`) AND (datum_ukoncenia=0) LEFT OUTER JOIN `pracovnyPomerTyp` `typ_pracovneho_pomeru` ON (`typ_pracovneho_pomeru`.`id`=`uziv_2`.`typ_pracovneho_pomeru`)

I can't figure out where's the problem, the query does not use the unique aliases, it uses the first ('uziv_2') only. I would be really appriciated if someone could help me, thank you.

**Comment by PatchRanger** [comment]
> the query does not use the unique aliases, it uses the first ('uziv_2') only

Confirmed, it always takes the first alias - so the resulting select is not such as expected.

**Comment by PatchRanger** [comment]
Here is related pull-request: #3850.

**Comment by samdark** [comment]
Won't be fixed in 1.1.
